bin/php-compat-cache.php

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * Scan a WordPress.org theme or plugin for compatibilty and cache results.
+ *
+ * ## OPTIONS
+ *
+ * <type>
+ * : Whether this is a plugin or theme from WordPress.org
+ * ---
+ * options:
+ *   - plugin
+ * ---
+ *
+ * <name>
+ * : Plugin or theme name (aka slug).
+ *
+ * <cache-dir>
+ * : Path to the cache directory.
+ *
+ * [--prior_versions=<count>]
+ * : How many prior versions to scan.
+ * ---
+ * default: 10
+ * ---
+ *
+ * @when before_wp_load
+ */
+WP_CLI::add_command( 'php-compat-cache', function( $args, $assoc_args ){
+	list( $type, $name, $cache_dir ) = $args;
+
+	if ( empty( $cache_dir ) || ! is_dir( $cache_dir ) ) {
+		WP_CLI::error( "Please make sure the cache directory exists before proceeding." );
+	}
+	$cache_dir = Utils\trailingslashit( realpath( $cache_dir ) );
+
+	exec( 'mkdir -p ' . escapeshellarg( $cache_dir . $name ), $output, $code );
+	if ( 0 !== $code ) {
+		WP_CLI::error( 'Failed to create target cache dir: '. $cache_dir . $name );
+	}
+
+	$versions = array();
+	if ( 'plugin' === $type ) {
+		$request_url = sprintf( 'https://api.wordpress.org/plugins/info/1.0/%s.json', $name );
+		$response = Utils\http_request( 'GET', $request_url );
+		if ( 200 !== $response->status_code ) {
+			WP_CLI::error( "{$response->status_code} HTTP response" );
+		}
+		if ( empty( $response->body ) || 'null' === $response->body ) {
+			WP_CLI::error( 'Plugin not found.' );
+		}
+		$plugin_data = json_decode( $response->body, true );
+		if ( empty( $plugin_data['versions'] ) ) {
+			WP_CLI::error( 'No plugin versions found.' );
+		}
+		// Versions are returned lowest to highest
+		$versions = array_reverse( $plugin_data['versions'], true );
+		unset( $versions['trunk'] );
+	}
+
+	$base_tmp_dir = Utils\get_temp_dir();
+	$prepare_dir = Utils\trailingslashit( $base_tmp_dir ) . $name . '-php-compat-cache/';
+
+	$prior_versions = $assoc_args['prior_versions'] < count( $versions ) ? $assoc_args['prior_versions'] : count( $versions );
+	WP_CLI::log( 'Scanning prior ' . $prior_versions . ' of ' . count( $versions ) . ' total ' . $name . ' versions' );
+	$php_versions = array(
+		'5.2',
+		'5.3',
+		'5.4',
+		'5.6',
+		'7.0',
+		'7.1',
+		'7.2',
+	);
+	$i = 0;
+	foreach ( array_slice( $versions, 0, $prior_versions ) as $plugin_version => $download_link ) {
+		$i++;
+		if ( is_dir( $prepare_dir ) ) {
+			exec( 'rm -r ' . escapeshellarg( $prepare_dir ), $output, $code );
+			if ( 0 !== $code ) {
+				WP_CLI::error( 'Failed to remove prepare dir: '. $prepare_dir );
+			}
+		}
+		exec( 'mkdir -p ' . escapeshellarg( $prepare_dir ), $output, $code );
+		if ( 0 !== $code ) {
+			WP_CLI::error( 'Failed to create prepare dir: '. $prepare_dir );
+		}
+		$download_fname = basename( $download_link );
+		WP_CLI::log( 'Downloading ' . $name . ' version ' . $plugin_version . ' (' . $i . '/' . $prior_versions . ')' );
+		exec( 'wget -q -O ' . escapeshellarg( $prepare_dir . $download_fname ) . ' ' . escapeshellarg( $download_link ), $output, $code );
```

Comment:
What if `wget` isn't installed? Should this use a PHP function instead to fetch the file, even `file_get_contents()`?

Reply:
> What if `wget` isn't installed?

It's installed on my local machine :) Which is probably the only place this will be run.

Reply:
Fair enough 😄